crates/builder/src/input/dotgraph.rs

```diff
@@ -265,23 +203,332 @@ where
+    }
+}
+
+impl<Label, G, NI> From<&G> for NeighborLabelFrequencies<Label, NI>
+where
+    NI: Idx,
+    Label: Idx + Hash,
+    G: Graph<NI>
+        + UndirectedNeighbors<NI>
+        + UndirectedDegrees<NI>
+        + NodeValues<NI, Label>
+        + Send
+        + Sync,
+{
+    fn from(graph: &G) -> Self {
+        let mut frequencies = Vec::with_capacity(graph.node_count().index());
+
+        (0..graph.node_count().index())
+            .into_par_iter()
+            .map(|node| {
+                let mut frequency = FxHashMap::<Label, usize>::default();
+
+                for &target in graph.neighbors(NI::new(node)) {
+                    let target_label = graph.node_value(target);
+                    let count = frequency.entry(*target_label).or_insert(0);
+                    *count += 1;
+                }
+
+                frequency
+            })
+            .collect_into_vec(&mut frequencies);
+
+        Self {
+            frequencies,
+            phantom: PhantomData::default(),
+        }
+    }
+}
+
+pub struct NodeLabelIndex<Label, NI>(Csr<Label, NI, ()>)
+where
+    NI: Idx,
+    Label: Idx;
+
+impl<Label, NI> NodeLabelIndex<Label, NI>
+where
+    NI: Idx,
+    Label: Idx,
+{
+    pub fn from_stats<F>(node_count: NI, label_stats: LabelStats<NI, Label>, label_func: F) -> Self
+    where
+        Label: Hash,
+        F: Fn(NI) -> Label + Send + Sync,
+    {
+        (node_count, label_stats, label_func).into()
+    }
+
+    pub fn nodes(&self, label: Label) -> &[NI] {
+        self.0.targets(label)
+    }
+}
+
+impl<Label, NI, F> From<(NI, LabelStats<NI, Label>, F)> for NodeLabelIndex<Label, NI>
+where
+    NI: Idx,
+    Label: Idx + Hash,
+    F: Fn(NI) -> Label + Send + Sync,
+{
+    fn from((node_count, label_stats, label_func): (NI, LabelStats<NI, Label>, F)) -> Self {
+        let LabelStats {
+            label_count,
+            max_label,
+            label_frequency,
+            ..
+        } = label_stats;
+        // Prefix sum: We insert the offset entries one index to the right and
+        // increment the offset of the next label during insert. That way we'll
+        // end up with the correct offsets after inserting into `nodes` in the
+        // next loop.
+        let mut offsets = Vec::with_capacity(label_count.index() + 1);
+        offsets.push(Label::zero());
+
+        let mut total = Label::zero();
+        for label in Label::zero().range_inclusive(max_label) {
+            offsets.push(total);
+            total += Label::new(*label_frequency.get(&label).unwrap_or(&0));
+        }
+
+        let mut offsets = ManuallyDrop::new(offsets);
+        let (ptr, len, cap) = (offsets.as_mut_ptr(), offsets.len(), offsets.capacity());
+
+        // SAFETY: Label and Label::Atomic have the same memory layout
+        let offsets = unsafe {
+            let ptr = ptr as *mut Atomic<Label>;
+            Vec::from_raw_parts(ptr, len, cap)
+        };
```

Comment:
This code looks a bit familiar :)
I think, however, we could improve the understanding of this totally-safe-transmute if we put this in its own block:

```suggestion
        let offsets = {
            let mut offsets = ManuallyDrop::new(offsets);
            let (ptr, len, cap) = (offsets.as_mut_ptr(), offsets.len(), offsets.capacity());

            // SAFETY: Label and Label::Atomic have the same memory layout
            unsafe {
            let ptr = ptr as *mut Atomic<Label>;
            Vec::from_raw_parts(ptr, len, cap)
            }
        }; 
```

Reply:
I like it :)